csrc/quickreduce/quick_reduce.h

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+#include <vector>
+#include <hip/hip_runtime.h>
+#include "quick_reduce_impl.cuh"
+
+#define HIP_CHECK(err)                                                     \
+  do {                                                                     \
+    hipError_t err_ = (err);                                               \
+    if (err_ != hipSuccess) {                                              \
+      std::printf("HIP error %d at %s:%d. %s\n", err_, __FILE__, __LINE__, \
+                  hipGetErrorString(err_));                                \
+      throw std::runtime_error("HIP error");                               \
+    }                                                                      \
+  } while (0)
+
+namespace quickreduce {
+using fptr_t = int64_t;
+static_assert(sizeof(void*) == sizeof(fptr_t));
+
+static constexpr int kOneShotAllreduceMaxElemsWorldSize2 = 8192 * 12;
+static constexpr int kOneShotAllreduceMaxElemsWorldSize4 = 8192 * 8;
+static constexpr int kOneShotAllreduceMaxElemsWorldSize8 = 8192 * 4;
+static constexpr long kOneShotAllreduceMaxSize =
+    std::max(kOneShotAllreduceMaxElemsWorldSize2 * 2,
+             std::max(kOneShotAllreduceMaxElemsWorldSize4 * 4,
+                      kOneShotAllreduceMaxElemsWorldSize8 * 8)) *
+    sizeof(half);
```

Comment:
nit: since all of the operands are 32-bit ints, then all of the arithmetic is performed in 32-bit. The conversion to `long` happens afterwards, so there's no reason for `kOneShotAllreduceMaxSize` to be a long.